### PR TITLE
Updating to clean up dataframe

### PR DESCRIPTION
**Pull request by VarshaMAthreya**
Changes made: 
1. Missing hearing thresholds - not tested/not recorded - marked as NaN 
2. "NR" marked as 121 dB - need to clean this further. 
3. The default for hearing loss classifier determining the type of hearing loss changed to 'unknown' instead of 'normal'. There are five categories in the classifier now -- 'Normal', 'Conductive', 'SNHL', 'Mixed', _'Unknown'_
4. In cases of symmetric HL with missing BC threshold in one ear, it's copied from the other ear to identify the type of HL - need to clean this further. 
5. SII calculated for both ears, with default changed from right ear to 'any ear' with available AC thresholds.

### CODE REVIEW

File: .vs/PythonSettings.json
Comment (MalcolmSlaney) [inline .vs/PythonSettings.json:1]: We should probably add .vs/* to the .gitignore file.  We don't want to check in all these VS files.  OK?

File: __pycache__/clusters.cpython-311.pyc
Comment (MalcolmSlaney) [inline __pycache__/clusters.cpython-311.pyc:1]: Likewise, add __pycache__ to .gitignore.,

File: clusters.py [hunk 1]
```diff
@@ -16,21 +16,18 @@
...
 
 import gspread
-import google.auth as auth # authenticatiing to google
+import google.auth as auth # authenticating to google
 
 #Global variables/parameters
 
-default_cluster_dir = ('/content/drive/MyDrive/Stanford Audiology Models/'
-                       'Colab Notebooks/')
+default_cluster_dir = ('/content/drive/MyDrive/STAR_Audiology_Model/')
```
Comment (MalcolmSlaney) [inline clusters.py:23]: We should probably talk and agree on a default that makes sense for both of us.  Not sure what that might be... perhaps the current directory (and then we can put a link into the place we each use.)?

File: clusters.py [hunk 2]
```diff
@@ -16,21 +16,18 @@
...
 
 """
 A change is required in the spreadsheet_path. Open the spreadsheet mfb raw audio
 and copy the link from the address bar. The gid code in the url is unique for
 each of the user.
 """
-spreadsheet_path_v1 = ('https://docs.google.com/spreadsheets/d/'
-                       '119_-qrfzGXwV1YBUJdBzvtAQTZnl-xwN7hD9FK5SWfU/'
-                       'edit#gid=84023254')
+spreadsheet_path_v1 = ('https://docs.google.com/spreadsheets/d/1M3pM45Qb2z2pZ_liVssCpOwCVb8fHtaeu1kkiomzxWw/edit#gid=0')
```
Comment (MalcolmSlaney) [inline clusters.py:30]: Did this change?  It's unfortunate that the path is not readable.  :-(

**Review by MalcolmSlaney — CHANGES_REQUESTED** [review]
Looking good. 
Also the test fails.. not sure why.  Do you know how to run it?

Will review the rest later today.

File: clusters.py (hunk 2 above)
Comment (VarshaMAthreya) [reply clusters.py:30]: yeah, I tried with a blank GSheet. We can ignore this change

File: .gitignore
Comment (MalcolmSlaney) [inline .gitignore:1]: Not sure why this change to .gitignore isn't removing these files.  Do you need to discard/revert this directory so that new commits don't have the .vs files?  I'm confused.

File: clean_data.py [hunk 3]
```diff
@@ -43,30 +43,40 @@ def replace_mrn(df, key: str = 'ReplaceMe') -> pd.DataFrame:
 
 def classify_hearing_loss(df: pd.DataFrame):
   # Assign result to a temporary dataframe, since classifier creates lots of 
-  # temporary column names.
+  # temporary column names.
```
Comment (MalcolmSlaney) [inline clean_data.py:46]: Looks like some extra spaces at the end of the line cause the diff to fail.  Can you remove them or revert this line?

File: clean_data.py [hunk 4]
```diff
@@ -23,7 +23,7 @@
 from scipy import interpolate
 
 import clusters
-from speech_intelligibility_index import sii
+from speech_intelligibility_index import sii   #Need to pip download
```
Comment (MalcolmSlaney) [inline clean_data.py:26]: More accurately, it should be pip install.

File: clean_data.py [hunk 5]
```diff
@@ -43,30 +43,40 @@ def replace_mrn(df, key: str = 'ReplaceMe') -> pd.DataFrame:
...
   orig_df = df.copy()
   new_df = clusters.HLossClassifier(df)
   df = orig_df
-  df['R_Type_HL_Mod'] = new_df['R_Type_HL_Mod']
-  df['R_Type_HL_HF'] = new_df['R_Type_HL_HF']
-  df['R_Type_HL_All'] = new_df['R_Type_HL_All']
+  
+  # Specify columns related to bone conduction, hearing loss types, and PTA to be transferred
+  bc_columns = ['RBone250', 'RBone500', 'RBone1000','RBone2000', 'RBone4000', 
+              'LBone250', 'LBone500', 'LBone1000', 'LBone2000', 'LBone4000']
```
Comment (MalcolmSlaney) [inline clean_data.py:54]: Looks like this line isn't properly indented.  Add two spaces?  Similarly below for hl_type_columns.

File: clean_data.py [hunk 6]
```diff
@@ -43,30 +43,40 @@ def replace_mrn(df, key: str = 'ReplaceMe') -> pd.DataFrame:
...
-# From: https://colab.research.google.com/drive/1bAnDpUUx5-BYkL3l3ukNVVnsLzEhy-ds#scrollTo=v6Dd6A2-UWG3
+# From: Google SII Colab Demo: https://colab.research.google.com/drive/1bAnDpUUx5-BYkL3l3ukNVVnsLzEhy-ds#scrollTo=v6Dd6A2-UWG3
 
 default_audiogram_freqs = [125, 250, 500, 750, 1000, 1500, 2000, 3000, 4000, 6000, 8000]
 
 def sii_from_audiogram(
     audiogram_samples: Union[List[float], np.ndarray],
     audiogram_freqs: Union[List[float], np.ndarray] = default_audiogram_freqs) -> float:
-    """Compute the speech intelligibility index from an audiogram
+    """Compute the speech intelligibility index from an audiogram -- This is from the SII module
```
Comment (MalcolmSlaney) [inline clean_data.py:79]: I'd drop the the -- comment.  It should be clear from the sii.sii call, and the import above.

File: clean_data.py [hunk 7]
```diff
@@ -91,54 +101,34 @@ def sii_from_audiogram(
     return sii.sii(ssl=ssl, nsl=nsl, hearing_threshold=critical_band_hl)
 
 
-def sii_from_df(df, ear='R'):
+def sii_from_df(df, ear):
```
Comment (MalcolmSlaney) [inline clean_data.py:104]: Can you add types to this call?   Ear should be a string, but even better would be to make it a literal, which I just discovered.  That way its clear what the proper values are.  This returns a float, right?

https://stackoverflow.com/questions/64057146/use-type-checking-to-check-that-string-is-of-certain-values-only

File: clean_data.py [hunk 8]
```diff
@@ -91,54 +101,34 @@ def sii_from_audiogram(
...
-
+    ear_sii = np.nan
+  return ear_sii
 
 def calculate_all_sii(df: pd.DataFrame):
   df['R_SII'] = df.apply(lambda df: sii_from_df(df, 'R'), axis=1)
   df['L_SII'] = df.apply(lambda df: sii_from_df(df, 'L'), axis=1)
   return df
 
-
```
Comment (MalcolmSlaney) [inline clean_data.py:118]: Just curious, why did this code go away?  I don't see it computed somewhere else.  Is it?

File: clean_data.py [hunk 9]
```diff
@@ -162,8 +152,7 @@ def label_duplicates(df: pd.DataFrame) -> pd.DataFrame:
   df[duplicate_name] = False
   df.loc[df['Patients::HMAC'].duplicated(keep=False), 'MultipleVisits'] = True
   return df
-
-
+
```
Comment (MalcolmSlaney) [inline clean_data.py:155]: Usually only one blank line *inside* functions.  (But two between functions.)

File: clean_data.py [hunk 10]
```diff
@@ -188,21 +179,25 @@ def main(argv):
       for line in fp:
         line = line.strip()
         print(create_hmac(line, FLAGS.hmac_key, hash_type=hashlib.sha3_384))
-  return
+    return
 
   df = pd.read_csv(FLAGS.input,
                    on_bad_lines='warn',
-                   na_values=['NR', '\n',
+                   na_values=['', '\n',
```
Comment (MalcolmSlaney) [inline clean_data.py:186]: You should add a comment, here, or in the function definition, stating what the policy is.

File: merge_audio_icd.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
```
Comment (MalcolmSlaney) [inline merge_audio_icd.py:1]: Can you not include this file in this PR?  It's not ready yet, and it's really something completely different.  That will keep the history cleaner.

**Review by MalcolmSlaney — CHANGES_REQUESTED** [review]
Sorry for the tardy response.

Looks like the tests are still failing.  Do you need help with that?  Can you run the tests on your machine before checking in your changes?

File: .gitignore
Comment (MalcolmSlaney) [reply .gitignore:1]: The test file mentions a new classification, 'Normal', but I don't see that label defined anywhere in the checked in code, in particular clusters.py.

Did you commit that change?

**Review by MalcolmSlaney — COMMENTED** [review]
Test failure question...

File: .gitignore
Comment (VarshaMAthreya) [reply .gitignore:1]: Yeah, I can see 'Normal' in clusters.py - Lines 288, 302, 315, 329, 343, 356.

**Comment by MalcolmSlaney** [comment]
For reasons I don’t understand.. the entire clusters.py file changed!  That’s not good.. that makes it too hard to review what changed and why.

Can you revert and apply your changes to a clean copy?

- Malcolm
